include/fiction/algorithms/iter/bdl_input_iterator.hpp

```diff
@@ -0,0 +1,317 @@
+//
+// Created by marcel on 24.07.23.
+//
+
+#ifndef FICTION_BDL_INPUT_ITERATOR_HPP
+#define FICTION_BDL_INPUT_ITERATOR_HPP
+
+#include "fiction/algorithms/simulation/sidb/detect_bdl_pairs.hpp"
+#include "fiction/types.hpp"
+
+#include <cstdint>
+#include <iterator>
+#include <vector>
+
+namespace fiction
+{
+
+/**
+ * Iterator that iterates over all possible input states of a BDL layout. There are \f$ 2^n \f$ possible input states
+ * for an \f$n\f$-input BDL layout, each with a unique input index. The input index is interpreted as a binary number,
+ * where the \f$i\f$-th bit represents the input state of the \f$i\f$-th input BDL pair. If the bit is `1`, the lower
+ * BDL dot is set and the upper BDL dot removed. If the bit is `0`, the upper BDL dot is removed and the lower BDL dot
+ * set. To this end, the iterator alters the given layout. The state enumeration wraps around, i.e. after the last
+ * possible input state, the first input state is set again.
+ *
+ * The iterator satisfies the requirements of `LegacyRandomAccessIterator` and can be used in iterator-based `for`
+ * loops.
+ *
+ * @tparam Lyt SiDB cell-level layout type.
+ */
+template <typename Lyt>
+class bdl_input_iterator
+{
+  public:
+    /**
+     * Standard constructor. It alters the layout to set the first input state, which assigns binary `0` to all input
+     * BDL pairs.
+     *
+     * @param lyt The SiDB BDL layout to iterate over.
+     * @param params Parameters for the BDL pair detection.
+     */
+    explicit bdl_input_iterator(Lyt& lyt, const detect_bdl_pairs_params& params = {}) noexcept :
+            layout{lyt},
+            input_pairs{detect_bdl_pairs(lyt, sidb_technology::cell_type::INPUT, params)},
+            num_inputs{static_cast<uint8_t>(input_pairs.size())}
+    {
+        static_assert(is_cell_level_layout_v<Lyt>, "Lyt is not a cell-level layout");
+        static_assert(has_sidb_technology_v<Lyt>, "Lyt is not an SiDB layout");
+        static_assert(has_siqad_coord_v<Lyt>, "Lyt is not based on SiQAD coordinates");
+
+        set_all_inputs();
+    }
+    /**
+     * Dereference operator. Returns a reference to the layout with the current input state.
+     *
+     * @return Reference to the current layout.
+     */
+    [[nodiscard]] Lyt& operator*() const noexcept
+    {
+        return layout;
+    }
+    /**
+     * Prefix increment operator. Sets the next input state.
+     *
+     * @return Reference to `this`.
+     */
+    bdl_input_iterator& operator++() noexcept
+    {
+        ++current_input_index;
+
+        set_all_inputs();
+
+        return *this;
+    }
+    /**
+     * Postfix increment operator. Sets the next input state.
+     *
+     * @return Copy of `this` before incrementing.
+     */
+    bdl_input_iterator operator++(int) noexcept
```

Comment:
Is there a reason why ``uint64_t`` is not used?

Reply:
Because that's how the C++ Standard defines these operators: https://en.cppreference.com/w/cpp/language/operators 

It's the same answer for all of these comments :smile: